src/common/hist_util.cc

```diff
@@ -350,12 +404,13 @@ void DenseCuts::Init
 
 void GHistIndexMatrix::Init(DMatrix* p_fmat, int max_num_bins) {
   cut.Build(p_fmat, max_num_bins);
+  max_num_bins_ = max_num_bins;
   const int32_t nthread = omp_get_max_threads();
   const uint32_t nbins = cut.Ptrs().back();
   hit_count.resize(nbins, 0);
   hit_count_tloc_.resize(nthread * nbins, 0);
 
-
+this->p_fmat_ = p_fmat;
```

Comment:
Where is this variable  used?

Reply:
It used for `ColumnMatrix` initialization to improve performance, it allows to reduce complexity from O(n_features) to O(1)
by replacing 
```
        auto iter = std::upper_bound(gmat.cut.Ptrs().cbegin() + fid,
                                     gmat.cut.Ptrs().cend(), bin_id);
        fid = std::distance(gmat.cut.Ptrs().cbegin(), iter) - 1;
```
to
`fid = inst[j].index;`
column_matrix.h:274 (in )